ssm/lds.py

```diff
@@ -681,51 +808,75 @@ def _objective(params, itr):
+                    log_likes = self.dynamics.log_likelihoods(x, input, x_mask, tag)
+                    log_likes += self.emissions.log_likelihoods(data, input, mask, tag, x)
+
+                    # Compute the expected log probability
+                    exp_log_joint += np.sum(Ez[0] * log_pi0)
+                    exp_log_joint += np.sum(Ezzp1 * log_Ps)
+                    exp_log_joint += np.sum(Ez * log_likes)
+            return exp_log_joint / n_samples
+
+        def estimate_entropy(n_samples):
+            entropy = 0.0
+            for sample in range(n_samples):
+                continuous_samples = variational_posterior.\
+                    sample_continuous_states()
+                entropy += variational_posterior.entropy(sample=continuous_samples)
+            return entropy / n_samples
+
+        # If expectations are given we can do an exact entropy calculation.
+        # Otherwise, entropy is estimated via samples.
+        elbo = estimate_expected_log_joint(n_samples)
+        if expectations is None:
+            elbo += estimate_entropy(n_samples)
+        else:
+            elbo += variational_posterior.entropy(expectations=expectations)
 
-    def _laplace_em_elbo(self, variational_posterior, datas, inputs, masks, tags, n_samples=1):
-
-        elbo = 0.0
-        for sample in range(n_samples):
-
-            # sample continuous states
-            continuous_samples = variational_posterior.sample_continuous_states()
-            discrete_expectations = variational_posterior.mean_discrete_states
-
-            # log p(theta)
-            elbo += self.log_prior()
-
-            for x, (Ez, Ezzp1, _), data, input, mask, tag in \
-                zip(continuous_samples, discrete_expectations, datas, inputs, masks, tags):
-
-                # The "mask" for x is all ones
-                x_mask = np.ones_like(x, dtype=bool)
-                log_pi0 = self.init_state_distn.log_initial_state_distn
-                log_Ps = self.transitions.log_transition_matrices(x, input, x_mask, tag)
-                log_likes = self.dynamics.log_likelihoods(x, input, x_mask, tag)
-                log_likes += self.emissions.log_likelihoods(data, input, mask, tag, x)
-
-                # Compute the expected log probability
-                elbo += np.sum(Ez[0] * log_pi0)
-                elbo += np.sum(Ezzp1 * log_Ps)
-                elbo += np.sum(Ez * log_likes)
+        return elbo
 
-            # add entropy of variational posterior
-            elbo += variational_posterior.entropy(continuous_samples)
 
-        return elbo / n_samples
 
     def _fit_laplace_em(self, variational_posterior, datas,
                         inputs=None, masks=None, tags=None,
                         num_iters=100,
                         num_samples=1,
-                        continuous_optimizer="newton",
+                        continuous_optimizer="lbfgs",
```

Comment:
Do you still encounter issues when using `newton`? If those have been identified and fixed, I think it would be nice to make `newton` the default because it is substantially faster in my experience.

Reply:
We should talk more about this.  Ben did some profiling and found that LBFGS was comparable, but it could very well be model-specific.  What models were you comparing?  Poisson rSLDS?

Reply:
Yes I notice the difference for the decision making Poisson rSLDS models. I didn't do a thorough comparison, but just noticed inference was taking longer than usual until I switched to newton. It is possible this is an unusual case.

Reply:
I think it is important to note I am typically fitting to datasets with on the order of 100 time series. So in that case computational differences between the algorithms are magnified by ~100x.